PSA: Use Mbed Crypto as a submodule

Patater opened the pull request:
## Description
Update Makefile and CMake build systems to support building and using libmbedcrypto as built from a submodule when `USE_CRYPTO_SUBMODULE` is defined.

This is intended to be merged after the next LTS release.

### Design principles
- Minimize changes to the existing Mbed TLS repository
- New build option only changes behavior of build system when used (USE_CRYPTO_SUBMODULE)
- Crypto still maintained in mbedtls for now, and only not used when USE_CRYPTO_SUBMODULE is defined
- Why not use crypto from mbedtls-psa yet? We don't want to cause build system issues.

### How to use
- `make USE_CRYPTO_SUBMODULE:=1 test` -> Build and test both Mbed TLS and Mbed Crypto, using libmbedcrypto from the "crypto" submodule
- `cmake -DUSE_CRYPTO_SUBMODULE=1 ..` -> Create build files that build libmbedcrypto in the "crypto" submodule and use it

### Funny things
- version.c is only included in libmbedcrypto
- oid.c is in libmbedcrypto
- error.c is in libmbedcrypto, and must know about SSL errors
- A run of all.sh with `USE_CRYPTO_SUBMODULE:=1` may not work yet

## Status
**IN DEVELOPMENT**

Don't merge. This is waiting for the Mbed Crypto repo to be ready and go public.

## Requires Backporting
NO

## Migrations
There are no intended API or ABI changes when the default configuration is used. When `USE_CRYPTO_SUBMODULE` is used, the only intended API change is for adding PSA APIs. These intentions have not yet been tested.

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated

Patater commented on visualc/VS2010/mbedTLS.vcxproj:
```diff
@@ -224,6 +224,11 @@
     <ClInclude Include="..\..\include\mbedtls\x509_crt.h" />
     <ClInclude Include="..\..\include\mbedtls\x509_csr.h" />
     <ClInclude Include="..\..\include\mbedtls\xtea.h" />
+    <ClInclude Include="..\..\crypto\include\psa\crypto.h" />
```
TODO: Don't modify the vcxproj files like this. Instead, add a new vcxproj file with crypto in it, modifying the vcxproj generating scripts to make it.

Patater replied:
Our tests want to run generate and see no changes, but to do that requires they know whether the submodule should be used or not. This will no longer be an issue when the submodule is always used, so I'd like to suggest we defer providing a crypto submodule using vcxproj until then.

Note this isn't just for PSA either, as when one uses the crypto submodule, they'd like to have the header files from the crypto submodule, not from mbedtls, whether or not they want PSA enabled.

Patater commented on tests/CMakeLists.txt:
```diff
@@ -49,73 +49,75 @@ if(MSVC)
+    add_test_suite(shax)
+    add_test_suite(timing)
+    add_test_suite(rsa)
+    add_test_suite(version)
```
TODO: get parity between CMake and Makefile testing. Currently, Makefile builds will test "version", "error", "debug" in both Mbed TLS and Mbed Crypto.

Patater replied:
OK

sbutcher-arm commented on CMakeLists.txt:
```diff
@@ -175,6 +181,9 @@ if(ENABLE_TESTING)
 
     add_subdirectory(tests)
+    if(ENABLE_PSA)
+        add_subdirectory(crypto/tests)
```
If we think of PSA Crypto as a separate§ library, when we run *our* tests, we may not care about the tests of the library as we might assume we're using a tested, proven version of the library. I think moving forwards, if we test Mbed TLS, we don't need to test crypto.

Patater replied:
Removed programs and tests, but kept library and include.

sbutcher-arm commented on CMakeLists.txt:
```diff
@@ -162,9 +161,16 @@ endif(ENABLE_ZLIB_SUPPORT)
 if(ENABLE_PROGRAMS)
     add_subdirectory(programs)
+    if(ENABLE_PSA)
+        add_subdirectory(crypto/programs)
```
If PSA Crypto is a library we happen to be using, do we care about building it's applications even if we don't use them? I don't think so. I think this can be removed.

Patater replied:
Removed programs and tests, but kept library and include.

sbutcher-arm commented on Makefile:
```diff
@@ -13,24 +13,51 @@ no_test: programs
 programs: lib
 	$(MAKE) -C programs
+ifdef ENABLE_PSA
+	$(MAKE) -C crypto/programs
```
As suggested above, I think we can remove this.

Patater replied:
Removed

sbutcher-arm commented on Makefile:
```diff
@@ -13,24 +13,51 @@ no_test: programs
 tests: lib
 	$(MAKE) -C tests
+ifdef ENABLE_PSA
+	$(MAKE) -C crypto/tests
```
As suggested above, I think we can remove this.

Patater replied:
Removed

sbutcher-arm commented on Makefile:
```diff
@@ -73,12 +112,18 @@ clean:
 check: lib tests
 	$(MAKE) -C tests check
+ifdef ENABLE_PSA
+	$(MAKE) -C crypto/tests check
```
Maybe we can remove this.

Patater replied:
Removed

sbutcher-arm commented on library/CMakeLists.txt:
```diff
@@ -140,42 +140,72 @@ elseif(USE_STATIC_MBEDTLS_LIBRARY)
-    set_target_properties(mbedcrypto PROPERTIES VERSION 2.13.1 SOVERSION 3)
-    target_link_libraries(mbedcrypto ${libs})
+    if(!ENABLE_PSA)
+        add_library(mbedcrypto SHARED ${src_crypto})
```
I'm not sure PSA Crypto should be redefining `mbedcrypto`. Better I think to make them separate targets, and one doesn't get built if you choose the other. Maybe that will make it too complicated though.

Patater replied:
I've made `mbedtls-psa` define `mbedcrypto`, and `mbedtls` not define `mbedcrypto`, if `USE_CRYPTO_SUBMODULE` is used.

If `USE_CRYPTO_SUBMODULE` is not used, `mbedtls` continues to define and use `mbedcrypto` and the `crypto` submodule is not referenced or used.

sbutcher-arm commented on programs/CMakeLists.txt:
```diff
@@ -1,8 +1,10 @@
 add_subdirectory(test)
-add_subdirectory(x509)
 add_subdirectory(util)
+add_subdirectory(ssl)
```
If this is alphabetical order, what alphabet are you using?

Patater replied:
The "put mbedtls specific things at the bottom" alphabet. I've made it actually alphabetical now.

sbutcher-arm commented on programs/pkey/CMakeLists.txt:
```diff
@@ -1,63 +1,71 @@
 
-add_executable(ecdh_curve25519 ecdh_curve25519.c)
-target_link_libraries(ecdh_curve25519 mbedtls)
+if(!ENABLE_PSA)
```
I don't think we need to do this. Why are we disabling them?

Patater replied:
If these programs are already part of the crypto submodule, why include them? We are not including crypto submodule tests, so we can exclude crypto submodule programs as well.

sbutcher-arm replied:
I think this is about transitioning to the Crypto module. As I see it, is goes in phases:

1. We add the submodule. There's duplication of some files. That's fine. Users get Mbed TLS as it was, and the option of the new stuff.
2. We remove the duplicates, and migrate stuff to the new module. We remove duplicated programs, test and library files.

We're working on 1. At the the moment if a program is duplicated, so what? At the moment if users want Mbed TLS they will expect the Mbed TLS programs to build. If they want PSA, they need to cd into the directory and build that directly. I think doing this before step 2 is confusing.

sbutcher-arm commented on programs/pkey/CMakeLists.txt:
```diff
@@ -1,63 +1,71 @@
-install(TARGETS dh_client dh_genprime dh_server key_app mpi_demo rsa_genkey rsa_sign rsa_verify rsa_encrypt rsa_decrypt pk_encrypt pk_decrypt pk_sign pk_verify gen_key
-        DESTINATION "bin"
-        PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)
+if(ENABLE_PSA)
```
I also don't thin we need to build the PSA programs.

Patater replied:
I agree. I'm considering crypto pkey programs to be everything except for `dh_client` `dh_server`. When `USE_CRYPTO_SUBMODULE` is set, only `dh_client` `dh_server` programs will be made from `mbedtls` the top level programs.

sbutcher-arm commented on programs/test/CMakeLists.txt:
```diff
@@ -13,8 +13,10 @@ endif(ENABLE_ZLIB_SUPPORT)
 
-add_executable(benchmark benchmark.c)
-target_link_libraries(benchmark ${libs})
+if(!ENABLE_PSA)
```
Again, I don't think this is necessary.

Patater replied:
Benchmark is only testing crypto. If the crypto submodule is enabled and used, why keep the top level benchmark program and not just use the crypto submodule one instead?

sbutcher-arm replied:
Because it's the Mbed TLS program using the alternative backend. I think enabling the submodule should change only the crypto library, nothing else. If it does before we move the files into the submodule it will be confusing.

sbutcher-arm commented on tests/CMakeLists.txt:
```diff
@@ -49,73 +49,75 @@ if(MSVC)
-add_test_suite(pkwrite)
-add_test_suite(poly1305)
-add_test_suite(shax)
+if(!ENABLE_PSA)
```
Why disable them?

Patater replied:
The crypto submodule will provide its own tests and use those. If the crypto submodule is used from the top level mbedtls, we don't really need to test the libmbedcrypto as built by the submodule with mbedtls's copy of the crypto tests: the crypto submodule tests are good enough.

sbutcher-arm replied:
Again - same point as before. At this stage, I think enabling the submodule should only change the supplied Crypto library functions. Not building files we haven't migrated yet, I think it confusing.

Patater commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "crypto"]
```
The intention is to eventually get even `mbedtls_*` crypto from here, not just PSA crypto. I've renamed the `ENABLE_PSA` build system option to `USE_CRYPTO_SUBMODULE` to better reflect the intent, which is to use libmbedcrypto as built by the crypto submodule instead of the `mbedtls` repo's copy.

sbutcher-arm replied:
My problem is we're moving from one Crypto interface to another, and I think it should be apparent. It's not just that we're *moving* the crypto files, but we're dropping one crypto API for another. The name blurs this.

Ultimately, it's your call.

Patater replied:
The API drop will happen after the submodule switch, not during or before. Switching to the submodule must not break API compatibility yet.

Patater commented on CMakeLists.txt:
```diff
@@ -9,6 +9,7 @@ option(USE_PKCS11_HELPER_LIBRARY "Build mbed TLS with the pkcs11-helper library.
 option(ENABLE_ZLIB_SUPPORT "Build mbed TLS with zlib library." OFF)
 
 option(ENABLE_PROGRAMS "Build mbed TLS programs." ON)
+option(ENABLE_PSA "Build and use PSA APIs." OFF)
```
Renamed to `USE_CRYPTO_SUBMODULE`

Patater commented on include/mbedtls/config.h:
```diff
@@ -1155,6 +1155,24 @@
+ * PSA APIs are subject to change. The implementation comes with less assurance
+ * and support than the rest of Mbed TLS.
+ *
+ * Module:  library/psa_crypto.c
```
Updated

Patater commented on include/mbedtls/config.h:
```diff
@@ -2584,6 +2602,22 @@
+ *
+ * Enable the Platform Security Architecture cryptography API.
+ *
+ * \warning This feature is experimental and available on an opt-in basis only.
```
Removed 'opt-in' wording.

Patater commented on include/mbedtls/config.h:
```diff
@@ -2584,6 +2602,22 @@
+ * PSA APIs are subject to change. The implementation comes with less assurance
+ * and support than the rest of Mbed TLS.
+ *
+ * Module:  library/psa_crypto.c
```
Fixed

Patater commented on programs/Makefile:
```diff
@@ -47,33 +51,61 @@ ifdef ZLIB
-	x509/cert_app$(EXEXT)		x509/crl_app$(EXEXT)		\
-	x509/cert_req$(EXEXT)		x509/cert_write$(EXEXT)		\
-	x509/req_app$(EXEXT)
+TLS_APPS := \
```
NOTE: We can revert the splitting of the APPS list for now, since the APPS list doesn't need to change.

Patater replied:
Done

mpg commented:
@Patater I don't think you meant to include https://github.com/ARMmbed/mbedtls/pull/2146/commits/caf72c866cb9c8cd1f7cea520190d4c185873275 in this PR. Can you remove it next time you push?

Patater commented:
@mpg I did mean to include it. The current version of the crypto submodule depends on that new function being present. Due to how header files are included from Mbed TLS and not from the submodule (so that config.h is pulled in from top level), the Mbed TLS and Mbed Crypto modules must have agreement in their headers. The change caf72c8 is required to get alignment in headers between the two repos.

mpg commented:
@Patater Ok, thanks for the clarification!

mpg commented:
@Patater I'm having with the submodule commit ID:
```
% git submodule update
fatal: reference is not a tree: a877122b30dcb661e784a0d1a4efa7a5d882ed67
Unable to checkout 'a877122b30dcb661e784a0d1a4efa7a5d882ed67' in submodule path 'crypto'
```
Also, I can't easily find that commit in the history of a clone of mbedtls-psa:
```
% git show a877122b30dcb661e784a0d1a4efa7a5d882ed67
fatal: bad object a877122b30dcb661e784a0d1a4efa7a5d882ed67
```
which is surprising as github is happy to show me that tree: https://github.com/ARMmbed/mbedtls-psa/tree/a877122b30dcb661e784a0d1a4efa7a5d882ed67

Any help would be welcome.

On a related note, I feel like to get a complete picture of how the build system works, I need to have a look at how things work on the other side of the fence - for example to satisfy myself that libmbedcrypto builds with the right `config.h` even from the submodule. I assume the correct PR to look at in order to satisfy my curiosity is https://github.com/ARMmbed/mbedtls-psa/pull/186 - I'm sating this here so you can correct me if I'm wrong, and for the benefit of other reviewers.

Patater commented on tests/scripts/all.sh:
```diff
@@ -574,6 +574,27 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # too slow for tests
+CC=gcc cmake -D USE_CRYPTO_SUBMODULE=1 -D CMAKE_BUILD_TYPE:String=Asan .
+make
+msg "test: main suites (USE_PSA)"
```
Fixed

Patater commented:
> I'd like to see a few words in Readme.md about the new build system option.

These few words were added to `crypto/README.md`. Using crypto from a submodule is experimental and not 100% supported yet, so didn't want to scare folks with updating the mbedtls README. TODO is mentioning this in the ChangeLog of mbedtls, though.

Patater commented on tests/scripts/all.sh:
```diff
@@ -574,6 +574,27 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+cp "$CONFIG_H" "$CONFIG_BAK"
+scripts/config.pl full
+scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # too slow for tests
+CC=gcc cmake -D USE_CRYPTO_SUBMODULE=1 -D CMAKE_BUILD_TYPE:String=Asan .
```
Changed what we test. We now test that libmbedcrypto was built with files from the crypto submodule, that the libmbedcrypto for the toplevel was not built, and that toplevel config.h is being used as we expect.

Patater commented:
Now using PR https://github.com/ARMmbed/mbedtls-psa/pull/188 which has a slightly different design (doesn't delete X.509 yet).

mpg commented:
> > I'd like to see a few words in Readme.md about the new build system option.
> 
> These few words were added to `crypto/README.md`. Using crypto from a submodule is experimental and not 100% supported yet, so didn't want to scare folks with updating the mbedtls README. TODO is mentioning this in the ChangeLog of mbedtls, though.

I agree with mentioning it only in ChangeLog while it's still experimental, and in the top-level readme only later - I just wanted a way for people to discover this, and the ChangeLog is fine for that.

mpg commented on tests/scripts/all.sh:
```diff
@@ -574,6 +574,43 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+cp "$CONFIG_H" "$CONFIG_BAK"
+scripts/config.pl full # enables md4 and submodule doesn't enable
+scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # too slow for tests
+make CC=gcc CFLAGS='-g' USE_CRYPTO_SUBMODULE=1
```
I'm getting the following error here:
```
  CC    psa_crypto.c
psa_crypto.c:41:24: fatal error: crypto_spe.h: No such file or directory
compilation terminated.
```

In case `MBEDTLS_PSA_CRYPTO_SPM` is not supposed to build yet or without additional platform-specific code, you might want to exclude it from what `config.pl full` enables, similarly to what's done [here](https://github.com/ARMmbed/mbedtls/commit/cd990a4961eebe9c10a90f14686d7e622f9c065d#diff-3bfa44e3db9ff239413e0258adf981c3)

Patater commented on tests/scripts/all.sh:
```diff
@@ -574,6 +574,43 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+msg "test: top-level libmbedcrypto wasn't built (USE_CRYPTO_SUBMODULE, cmake)"
+if_build_succeeded if [ -f library/libmbedcrypto.a ]; then false; fi
+msg "test: libmbedcrypto symbols are from crypto files (USE_CRYPTO_SUBMODULE, cmake)"
+if_build_succeeded objdump -g crypto/library/libmbedcrypto.a | grep 'crypto/library/aes.c'
```
Added

gilles-peskine-arm replied:
Just seeing this in passing… Note that the line `if_build_succeeded somecommand | grep stuff` doesn't work because it runs `grep stuff` with empty input if the build didn't succeed, which causes the script to exit even in keep-going mode. If you need to run more than a simple command after `if_build_succeeded`, you need to create a function.

(One Day we'll rewrite this script in a language with fewer quirks, like Python.)

Patater commented on tests/scripts/all.sh:
```diff
@@ -574,6 +574,43 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+CC=gcc cmake -D USE_CRYPTO_SUBMODULE=1 -D CMAKE_BUILD_TYPE=Debug
+make
+msg "test: top-level libmbedcrypto wasn't built (USE_CRYPTO_SUBMODULE, cmake)"
+if_build_succeeded if [ -f library/libmbedcrypto.a ]; then false; fi
```
Updated

Patater commented on tests/scripts/all.sh:
```diff
@@ -574,6 +574,43 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+cp "$CONFIG_H" "$CONFIG_BAK"
+scripts/config.pl full # enables md4 and submodule doesn't enable
+scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # too slow for tests
+CC=gcc cmake -D USE_CRYPTO_SUBMODULE=1 -D CMAKE_BUILD_TYPE=Debug
```
Sorry, this wasn't implemented yet. I've implemented it now.

mpg commented on include/mbedtls/config.h:
```diff
@@ -2586,6 +2604,22 @@
+ *
+ * Module:  crypto/library/psa_crypto.c
+ *
+ * Requires: MBEDTLS_CTR_DRBG_C, MBEDTLS_ENTROPY_C
```
Suggestion: we might want to mention that enabling this option has no effect unless you also use the build option `USE_CRYPTO_SUBMODULE`. Otherwise users might be surprised that they enabled this and don't get any of the promised APIs anyway.

Actually, we might even want the `USE_CRYPTO_SUBMODULE` build option and add `-DMBEDTLS_USE_CRYPTO_SUBMODULE` to `CFLAGS` and then in `check_config.h` error out if `MBEDTLS_PSA_CRYPTO_C` is defined but `MBEDTLS_USE_CRYPTO_SUBMODULE` isn't.

However, feel free to ignore that suggestion for this PR if you feel it would be feature creep - we can always add it in a subsequent PR.

Patater replied:
Updated to mention the build option.

Good idea about modifying CFLAGS. I think we'll make using the submodule the default option soon enough that it isn't investing too much time in this, though.

Patater commented:
Added some as yet unapplied fixup commits to update the submodule and the documentation for the PSA configuration options. Added a new commit for a ChangeLog entry.

mpg commented:
I just finished running tests for a (hopefully relevant) subset of `all.sh` (commented out the parts that had nothing to do with this option, except the sanity checks and and tests with the default config at the beginning). I'm happy to report it failed much less badly than last time, so we're definitely making progress.

The only failures I got were in `ssl-opt.sh` for the `USE_CRYPTO_SUBMODULE` builds. Both Make and CMake builds have the exact same set of 30 tests failing, out of the 583 that this script runs, so this doesn't look like random failures. Also, most of the failing tests passed in the `ssl-opt.sh` run without that build option (the others were skipped due to not using a full `config.h`).

I'm investigating these failures and will update here when I have more to say about their cause.

mpg commented:
Out of the 30 test failing, 2 seem to be known issues (the two "3d, gnutls server, fragmentation" tests are known to be fragile) that happened to manifest just in the `USE_CRYPTO_SUBMODULE` build, and the other 28 are caused by `mbedtls_strerror()` no working on SSL errors in that build.

I know it's a questionable choice to rely on `mbedtls_strerror()` in `ssl-opt.sh`, and one we might revisit at some point, but OTOH I don't think it's OK to break that function for our users now. In the long run we probably want to heavily rework that function or even replace it with something else entirely, but for now I'm not sure what we should do.

I didn't test `mbedtls_version_check_feature()` but it might be interesting to know if it's broken as well and have a plan about it too. I don't think there are any other similar problem beyond those two (and as you noted, `version.c` being only in `libmbedcrypto` so far, which I think is less of an issue in the very short term, but will need to be addressed later too).

Patater commented:
Thanks for the help with testing and reviewing so far. I'll see what we can do about `mbedtls_strerror()`.

Patater commented:
Updated crypto submodule with a new commit to revert the removal of SSL code. I think there is still a potential for errors here, as the crypto submodule errors.c might not have the same set of SSL errors and descriptions as mbedtls does.

Patater commented:
I have a passing `all.sh` for the `USE_CRYPTO_SUBMODULE` tests now, including `ssl-opt.sh`.

Patater commented:
Added a commit to fix a Makefile error, which led to too much getting built in some situations.

Patater commented:
This uses a submodule based on a new PR in Mbed Crypto now. https://github.com/ARMmbed/mbedtls-psa/pull/194

Also renamed INCLUDES to CRYPTO_INCLUDES to make the name more unique and informative.

Patater commented:
I've done a complete `all.sh` run and it passed.

Patater commented:
Add fixup commits to make sure the crypto submodule is initialized in all.sh. A previous idea to not require the crypto submodule to be initialized was rejected as we always want the test coverage.

Patater commented:
Regarding CI, yes, it needs to be updated to fetch submodules. That's on the list.

Regarding abi_check.py, that's a mistake on my part. As part of development testing, I was verifying that using the submodule doesn't affect the ABI. After the test, I intended to leave out the code that tested using the submodule, and I did leave that out, but I left in some code to make build the submodule. I've corrected this by adding a fixup commit.

Patater commented on include/mbedtls/config.h:
```diff
@@ -1155,6 +1155,24 @@
+ * Environment).
+ *
+ * \warning This feature is experimental and available for evaluation only. It
+ * is disabled by default. PSA APIs are subject to change. The implementation
```
Revised in a fixup commit

mpg commented:
> Regarding abi_check.py, that's a mistake on my part. As part of development testing, I was verifying that using the submodule doesn't affect the ABI. After the test, I intended to leave out the code that tested using the submodule, and I did leave that out, but I left in some code to make build the submodule. I've corrected this by adding a fixup commit.

Do you think you could make that a command-line option? We might also want to check later how using the submodule affects the ABI. Perhaps in a follow-up PR in order to keep this one nice and small.

Patater commented on tests/scripts/all.sh:
```diff
@@ -619,6 +619,29 @@ make CC=gcc USE_CRYPTO_SUBMODULE=1 test
+msg "test: submodule libmbedcrypto wasn't built (USE_CRYPTO_SUBMODULE, make)"
+if_build_succeeded not test -f crypto/library/libmbedcrypto.a
+msg "test: libmbedcrypto symbols are from library files (USE_CRYPTO_SUBMODULE, make)"
+if_build_succeeded objdump -g library/libmbedcrypto.a | grep -E 'library$' | grep -v 'crypto' > /dev/null
```
Fixed in fixup commit

Patater commented:
>> Regarding abi_check.py, that's a mistake on my part. As part of development testing, I was verifying that using the submodule doesn't affect the ABI. After the test, I intended to leave out the code that tested using the submodule, and I did leave that out, but I left in some code to make build the submodule. I've corrected this by adding a fixup commit.

>Do you think you could make that a command-line option? We might also want to check later how using the submodule affects the ABI. Perhaps in a follow-up PR in order to keep this one nice and small.

I think that's a good idea for a follow-up PR, yes. Maybe we can default to always using the submodule soon enough, though. Adding this is something to keep in mind if we can't always use the submodule for whatever reason.

gilles-peskine-arm commented on include/mbedtls/rsa.h:
```diff
@@ -403,6 +403,16 @@ void mbedtls_rsa_set_padding( mbedtls_rsa_context *ctx, int padding,
+ * \return         The length of the RSA modulus in bits.
+ *
+ */
+size_t mbedtls_rsa_get_bitlen( const mbedtls_rsa_context *ctx );
```
Unfortunately adding this function breaks existing ALT implementations since they don't provide it. Furthermore I don't like adding this extra function without adding a few more tests.

gilles-peskine-arm commented on Makefile:
```diff
@@ -28,7 +28,11 @@ install: no_test
 	cp -RP library/libmbedx509.*   $(DESTDIR)/lib
+ifdef USE_CRYPTO_SUBMODULE
+	$(MAKE) -C crypto install
+else
```
As far as I understand, `crypto/include/mbedtls/config.h` is only meant to be used when crypto is used on its own. Standalone crypto, at this point, is primarily a demonstrator for the PSA API, so it activates a different set of features from mbedtls. In particular, we want PSA to be present in the default crypto build, but not in the default mbedtls build.

We need to make sure that the build scripts don't mix them up and to make the documentation clear.

Patater replied:
We have tests to ensure the build scripts don't mix them up. The documentation for how to use the submodule is in the crypto/README.md file, but it needs updating to explain how config.h is handled.

Patater replied:
Added a new commit for handling `make install` with submodule properly.

sbutcher-arm replied:
For MbedOS 5.11 we will use only the mbed TLS config file - `include\mbedtls\config.h`. When the submodule is used on it's own, it will use its own `config.h` file.

In Mbed OS 5.12, we'll swap to two config.h files - one for TLS and X.509 in `include\mbedtls\config.h` and a separate one for Crypto.

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -1155,6 +1155,28 @@
+ * When MBEDTLS_PSA_CRYPTO_SPM is defined, the code is built for SPM (Secure
+ * Partition Manager) integration which separates the code into two parts: a
+ * NSPE (Non-Secure Process Environment) and an SPE (Secure Process
+ * Environment).
```
You can't build Crypto with SPM support using the code on this branch. `MBEDTLS_PSA_CRYPTO_SPM` is currently a hack. We should probably not include it in `config.h` in this way. And we need to make sure that it doesn't get included by any of the presets in `scripts/config.pl`.

Patater replied:
Both PSA config options added in this PR are also added to the exclusion list in `scripts/config.pl`.

Patater replied:
I'll remove the SPM option from the Mbed TLS config.h

gilles-peskine-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -1,7 +1,16 @@
+#PK utils: RSA, 511 bits
+#depends_on:MBEDTLS_RSA_C:MBEDTLS_GENPRIME
+#pk_utils:MBEDTLS_PK_RSA:511:64:"RSA"
+#
```
It would fail because mbedtls only supports the generation of even-sized RSA keys. We've just reworked the crypto branch to minimize non-PSA-related changes so this chunk of code will go away in the next iteration of this PR.

Patater commented on include/mbedtls/config.h:
```diff
@@ -2586,6 +2608,25 @@
+ *
+ * \warning This feature is experimental and available on an opt-in basis only.
+ * PSA APIs are subject to change at any time. The implementation comes with
+ * less assurance and support than the rest of Mbed TLS.
```
alpha, beta, or something is also vague, isn't it?

sbutcher-arm replied:
The maturity is defined by ATG. I'm not sure we can say anything better than it's an 'early release' at the moment.

Patater commented on tests/scripts/all.sh:
```diff
@@ -574,6 +587,66 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+if_build_succeeded not test -f crypto/library/libmbedcrypto.a
+msg "test: libmbedcrypto symbols are from library files (USE_CRYPTO_SUBMODULE, cmake)"
+if_build_succeeded objdump -g library/libmbedcrypto.a | grep -E 'library$' | not grep 'crypto' > /dev/null
+
```
Yes, that's needed.

mpg replied:
Note: hasn't been done yet, but as I commented earlier, I think it can wait for a follow-up PR.

Patater commented:
- Old branch at: https://github.com/Patater/mbedtls/tree/dev/Patater/use-crypto-submodule-1
- Rebased to apply fixups, with result at: https://github.com/Patater/mbedtls/tree/dev/Patater/use-crypto-submodule-2
- Rebased on latest development branch, and updating submodule, with result as this PR (or https://github.com/Patater/mbedtls/tree/dev/Patater/use-crypto-submodule-3)

mpg commented:
The latest commit doesn't build for me:
```
% make USE_CRYPTO_SUBMODULE=1 all                                                                                                                                  git|2146|
  CC    bignum.c
bignum.c: In function ‘mbedtls_mpi_gen_prime’:
bignum.c:2225:19: error: ‘MBEDTLS_MPI_GEN_PRIME_FLAG_LOW_ERR’ undeclared (first use in this function)
     if( ( flags & MBEDTLS_MPI_GEN_PRIME_FLAG_LOW_ERR ) == 0 )
                   ^
bignum.c:2225:19: note: each undeclared identifier is reported only once for each function it appears in
bignum.c:2256:23: error: ‘MBEDTLS_MPI_GEN_PRIME_FLAG_DH’ undeclared (first use in this function)
         if( ( flags & MBEDTLS_MPI_GEN_PRIME_FLAG_DH ) == 0 )
                       ^
Makefile:197: recipe for target 'bignum.o' failed
make[2]: *** [bignum.o] Error 1
Makefile:179: recipe for target 'libmbedcrypto.a' failed
make[1]: *** [libmbedcrypto.a] Error 2
Makefile:18: recipe for target 'lib' failed
make: *** [lib] Error 2
```
Probably some header differences between Mbed Crypto and Mbed TLS. Maybe we need a different version of the submodule?

Patater commented:
@mpg Yup, this won't work until after the next version of Mbed TLS ships.

mpg commented:
@Patater Good to know that the issue was known :)

Patater commented:
Rebased on mbedtls-2.14.0 and force pushed

Patater commented:
CI job needs updating to fetch submodules, before this will pass CI

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -88,6 +88,11 @@ elif [ -d library -a -d include -a -d tests ]; then :; else
     exit 1
 fi
 
+if ! [ -f crypto/Makefile ]; then
```
Non-blocker: I'd like an option `--no-submodule` to say “I don't care about the submodule, just test mbedtls.”

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -154,9 +159,12 @@ cleanup()
     command make clean
+    cd crypto
+    command make clean
+    cd ..
```
Non-blocker: this can break if `crypto` is a symlink. Replace these three lines by `make -C crypto clean`. Other places that use `cd crypto; …; cd ..` should be similarly protected.

mpg replied:
Note: An option (for commands other that actually require changing the current working directory) is to use a sub-shell: `(cd crypto; do_stuff; more_stuff)`.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -574,6 +587,66 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
 msg "test: compat.sh ARIA + ChachaPoly"
 if_build_succeeded env OPENSSL_CMD="$OPENSSL_NEXT" tests/compat.sh -e '^$' -f 'ARIA\|CHACHA'
 
+# USE_CRYPTO_SUBMODULE: check that the build works with CMake
```
Non-blocker: we should also test an out-of-tree build.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -574,6 +587,66 @@ if_build_succeeded env OPENSSL_CMD="$OPENSSL_LEGACY" GNUTLS_CLI="$GNUTLS_LEGACY_
+msg "test: main suites (USE_CRYPTO_SUBMODULE, cmake)"
+make test
+msg "test: ssl-opt.sh (USE_CRYPTO_SUBMODULE, cmake)"
+if_build_succeeded tests/ssl-opt.sh
```
Non-blocker: I think a subset of `compat.sh` would be a better test for exercising crypto features that `ssl-opt.sh`. `ssl-opt.sh` is mostly about exercising TLS features. Currently in PSA crypto we test that we aren't breaking the default mbedtls configuration with `compat.sh -m tls1_2 -V NO`. This applies to the other builds, or we could run different sets of tests with different build tools.

sbutcher-arm replied:
I don't think I agree. I see this as a system test of Mbed TLS - checking that Mbed TLS works with PSA Crypto and the submodule.

mpg commented:
> What is borderline blocking for me is the lack of documentation. There are no instructions on how to use mbedcrypto anywhere. I third (or more?) the request for a section in README.md.

I don't know if you're counting me in the people who requested a section in the Readme, but I've change my mind since then: I think so far the feature is only for evaluation by people who know they want to evaluate it, which so far mainly means ARM people. A confluence page will be created soon (there's a task on me to do that) with instructions.

I think that a section in the Readme will be useful later, but that now is not the best time.

mpg commented:
I've been testing the "make install" targets and I'm satisfied that they work as intended. More specifically, I've install with make and cmake both with an without the submodule, and then compared the resulting trees with the same build system with and without the submodule. The diff is as expected:
```
58a59
> psa
138a140,147
> 
> DESTDIR/include/psa:
> crypto.h
> crypto_driver.h
> crypto_extra.h
> crypto_platform.h
> crypto_sizes.h
> crypto_struct.h
```

I'll note however that no test for `make install` has been added to `all.sh` so far. I don't think that's an issue right now, considering (1) this was not tested before the submodule either and (2) in the very short term we don't care that much about `make install` as it's not used in Mbed OS. I think such tests could be added in a follow-up PR.

gilles-peskine-arm commented:
> I think that a section in the Readme will be useful later, but that now is not the best time.

I disagree: if nothing else writing documentation forces you to think about what you're doing.

For example, it's made me realize that when you build with `USE_CRYPTO_SUBMODULE`, it only enables the use of library files. It doesn't enable any tests from the submodule, or the sample/test/utility programs.

In particular, to make good use of the PSA API, you need to use Mbed Crypto as a standalone product, not as a submodule. That's the only way you'll get `programs/psa/*` and `test_suite_psa_crypto*`.

mpg commented:
This just in: a local run of `all.sh -k -r` completed successfully except for known issues addressed in #2209.

mpg commented:
> I disagree: if nothing else writing documentation forces you to think about what you're doing.

I agree with that part. Documentation is being written (well, will be as soon as I'm done reviewing this PR). I just disagree about placing the result in the Readme right now.

gilles-peskine-arm commented:
I've written a readme section, it's available for review at https://github.com/Patater/mbedtls/pull/1

mpg commented:
Removing "DO NOT MERGE" and "needs work" as I think they no longer apply - feel free to re-apply if I missed something or an issue comes up.

mpg commented:
> I've written a readme section, it's available for review at Patater#1

@gilles-peskine-arm Thanks!

Patater commented:
Updated submodule to point to a commit on the main branch, not a PR. Also cherry-picked Gilles's README updating commit into this PR.

sbutcher-arm commented:
The Jenkins CI release job, [#819](https://jenkins-internal.mbed.com/view/mbed-tls/job/mbedtls-release/819/), has now completed with some minor failures.
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
FAILED: 2
test: ssl-opt.sh (USE_CRYPTO_SUBMODULE, cmake): tests/ssl-opt.sh -> 2
test: ssl-opt.sh (USE_CRYPTO_SUBMODULE, make): tests/ssl-opt.sh -> 2
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Specifically, the two tests that failed were:
```
TLS proxy: 3d, gnutls server, fragmentation
DTLS proxy: 3d, gnutls server, fragmentation
```
These are susceptible to timing and are know to fail on heavily loaded test machines.

Therefore - the PR can be seen as good enough.

sbutcher-arm commented on include/mbedtls/config.h:
```diff
@@ -2590,6 +2590,25 @@
+ *
+ * \warning This feature is experimental and available on an opt-in basis only.
+ * PSA APIs are subject to change at any time. The implementation comes with
+ * less assurance and support than the rest of Mbed TLS.
```
[Not a blocker] I think we need to state that our long term plans are to make this the default.

sbutcher-arm commented on CMakeLists.txt:
```diff
@@ -9,6 +9,7 @@ option(USE_PKCS11_HELPER_LIBRARY "Build mbed TLS with the pkcs11-helper library.
 option(ENABLE_ZLIB_SUPPORT "Build mbed TLS with zlib library." OFF)
 
 option(ENABLE_PROGRAMS "Build mbed TLS programs." ON)
+option(USE_CRYPTO_SUBMODULE "Build and use libmbedcrypto from the crypto submodule." OFF)
```
[Not a blocker] I think we need to state that `USE_CRYPTO_SUBMODULE ` is a temporary option, and we will be moving the Crypto to the submodule and this will be not just the default, but the only option.

mpg commented on README.md:
```diff
@@ -158,6 +158,43 @@ Configurations
+1. Check out the `crypto` submodule and update it.
+
+        git submodule init crypto
+        git submodule update crypto
```
Well an option would be to have this text not in the Readme, but on a private page accessible only to users who can also access the private repo. We could then move it back to the Readme when the submodule repo becomes public.

sbutcher-arm commented:
This PR has now been merged to `development-psa` as the first PR to the branch, not `development` which was the base branch specified above. As such, this PR can now be closed.
